storage/cloudspanner/log_storage.go

```diff
@@ -577,44 +580,51 @@ func (tx *logTX) DequeueLeaves(ctx context.Context, limit int, cutoff time.Time)
 	// merkle hash of the entry.
 	now := time.Now().UTC()
 	cfg := tx.getLogStorageConfig()
-	timeBucket := int64(((now.Unix() + cfg.NumUnseqBuckets/2) % cfg.NumUnseqBuckets) << 8)
+	// Select a prefix that is likley to be on a different span server to spread load.
+	prefix := int64((((now.Unix() + cfg.NumUnseqBuckets/2) % cfg.NumUnseqBuckets) << 8))
 
 	// Choose a starting point in the merkle prefix range, and calculate the
 	// start/limit of the merkle range we'll dequeue from.
 	// It seems to be much better to tune for keeping this range small, and allow
 	// the signer to run multiple times per second than try to dequeue a large batch
 	// which spans a large number of merkle prefixes.
-	merklePrefix := rand.Int63n(256)
-	startBucket := timeBucket | merklePrefix
-	numMerkleBuckets := int64(256 * tx.ls.opts.DequeueAcrossMerkleBucketsRangeFraction)
-	merkleLimit := merklePrefix + numMerkleBuckets
-	if merkleLimit > 0xff {
-		merkleLimit = 0xff
-	}
-	limitBucket := timeBucket | merkleLimit
-
-	stmt := spanner.NewStatement(`
-			SELECT 
-			  Bucket, 
-			  QueueTimestampNanos, 
-			  MerkleLeafHash, 
-			  LeafIdentityHash
-			FROM 
-			  Unsequenced
-			WHERE 
-			  TreeID = @tree_id AND 
-			  Bucket >= @start_bucket AND 
-			  Bucket <= @limit_bucket
-			LIMIT 
-			  @max_num`)
-	stmt.Params["tree_id"] = tx.treeID
-	stmt.Params["start_bucket"] = startBucket
-	stmt.Params["limit_bucket"] = limitBucket
-	stmt.Params["max_num"] = limit
+	suffixStart := rand.Int63n(suffixBuckets)
+	suffixFraction := float64(cfg.NumMerkleBuckets) / float64(suffixBuckets)
+	if tx.ls.opts.DequeueAcrossMerkleBuckets {
+		suffixFraction = tx.ls.opts.DequeueAcrossMerkleBucketsRangeFraction
+	}
+	suffixEnd := suffixStart + int64(suffixBuckets*suffixFraction)
+
```

Comment:
I don't follow what this block would do or how it would be triggered?

Reply:
If `suffixFraction` is low enough, `suffixEnd` will be equal to `suffixStart`. Since below I suggest excluding `End`, it would result in an empty request. So +1 fixes that. As another alternative, you round `suffixEnd` up?

Reply:
Used `math.Ceil`